pkg/alerts/alertmanager/client.go

```diff
@@ -0,0 +1,118 @@
+package alertmanager
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"time"
+)
+
+const (
+	alertManagerPathPrefix = "/api/v2"
+	// the default context timeout for alert manager client
+	// feel free to change this value/set a corresponding env var if needed
+	defaultTimeOut = 10 * time.Second
+)
+
+// New is a constructor of AlertManagerClient
+func New() (*Client, error) {
+	if os.Getenv("ALERTMGR_URL") == "" {
+		return nil, fmt.Errorf("no alert manager url found")
+	}
+	return &Client{
+		Token:   os.Getenv("ALERTMGR_TOKEN"),
+		HostURL: os.Getenv("ALERTMGR_URL"),
+		Client:  new(http.Client),
+	}, nil
+}
+
+// Status builds the Status endpoint
+func (c *Client) Status() *StatusEndpoint {
+	return &StatusEndpoint{
+		Client: *c,
+	}
+}
+
+// Silences builds the Silences endpoint
+func (c *Client) Silences() *SilencesEndpoint {
+	return &SilencesEndpoint{
+		Client: *c,
+	}
+}
+
+// BuildURL builds the full URL for Status Endpoint
+func (s *StatusEndpoint) BuildURL() error {
+	url, err := url.Parse(s.HostURL)
+	if err != nil {
+		return err
+	}
+	url.Path = path.Join(alertManagerPathPrefix, "status")
+	s.FullURL = url.String()
+	return nil
+}
+
+// Get receives information about alert manager overall status
+func (s *StatusEndpoint) Get() (*StatusResponse, error) {
+	err := s.BuildURL()
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeOut)
+	defer cancel()
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, s.FullURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	response, err := s.Client.Client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	responseObject := new(StatusResponse)
+	err = json.NewDecoder(response.Body).Decode(responseObject)
+	if err != nil {
+		return nil, err
+	}
+	return responseObject, nil
+}
+
+// BuildURL builds the full URL for silences Endpoint
+func (s *SilencesEndpoint) BuildURL() error {
+	url, err := url.Parse(s.HostURL)
+	if err != nil {
+		return err
+	}
+	url.Path = path.Join(alertManagerPathPrefix, "silences")
+	s.FullURL = url.String()
+	return nil
+}
+
+// Get lists the silences
+func (s *SilencesEndpoint) Get() ([]GettableSilence, error) {
+	err := s.BuildURL()
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeOut)
+	defer cancel()
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, s.FullURL, nil)
+	if err != nil {
+		return nil, err
+	}
+	response, err := s.Client.Client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	responseObject := make([]GettableSilence, 0)
+	err = json.NewDecoder(response.Body).Decode(&responseObject)
+	if err != nil {
+		return nil, err
+	}
+	if err := ValidateStatus(responseObject); err != nil {
+		return nil, err
+	}
+	return responseObject, nil
+}
```

Comment:
I wanted to systematize the URL/endpoint construction and their methods to be split among them. 
So that so, when you call client.Silences().Get() just like in k8s client-go: client.CoreV1().Pods....

Reply:
Thanks for explaining that. It clarifies your goals with the structure. I think this should be in a comment in the code.